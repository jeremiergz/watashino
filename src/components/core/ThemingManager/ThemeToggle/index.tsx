import React from 'react';
import { useTheming } from '..';
import Button from '../../../common/Button';
import Flex from '../../../primitives/Flex';
import Text from '../../../primitives/Text';
import Moon from '../../../svg/Moon';
import Sun from '../../../svg/Sun';

const iconMapping = {
  dark: Moon,
  light: Sun,
};

const ThemeToggle = () => {
  const { theme, toggle } = useTheming();
  const ThemeIcon = iconMapping[theme.type];
  return (
    <Button onClick={() => toggle()} position="absolute" right={3} top={3} variant="outlined" width={72}>
      <Flex alignItems="center" justifyContent="space-between" width="100%">
        <ThemeIcon height={20} width={20} />
        <Text color="gray" fontFamily={theme.fonts.main} fontSize={14}>
          {theme.type}
        </Text>
      </Flex>
    </Button>
  );
};

ThemeToggle.displayName = 'ThemeToggle';

export default ThemeToggle;
