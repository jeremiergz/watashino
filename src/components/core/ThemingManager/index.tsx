import React, { createContext, PropsWithChildren, useContext, useState } from 'react';
import { ThemeProvider } from 'styled-components';
import { DarkTheme, LightTheme, Theme } from '../../../theme';

type ThemingManagerProps = PropsWithChildren<{}>;

const defaultMode = 'light';
const localStorageKey = 'theme.mode';

const ThemingManagerContext = createContext<{ mode: ThemeMode; theme: Theme; toggle: () => void }>({
  mode: defaultMode,
  theme: LightTheme,
  toggle: () => null,
});

function useTheming() {
  return useContext(ThemingManagerContext);
}

const ThemingManager = ({ children }: ThemingManagerProps) => {
  const isBrowser = typeof window !== 'undefined';
  const storedMode = (isBrowser && (localStorage.getItem(localStorageKey) as ThemeMode)) || defaultMode;
  const [mode, setMode] = useState<ThemeMode>(storedMode);
  const toggle = () => {
    const newMode = mode === 'light' ? 'dark' : 'light';
    setMode(newMode);
    if (isBrowser) localStorage.setItem(localStorageKey, newMode);
  };
  const theme = mode === 'light' ? LightTheme : DarkTheme;
  return (
    <ThemingManagerContext.Provider value={{ mode, theme, toggle }}>
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    </ThemingManagerContext.Provider>
  );
};

ThemingManager.displayName = 'ThemingManager';

export { useTheming };
export default ThemingManager;
