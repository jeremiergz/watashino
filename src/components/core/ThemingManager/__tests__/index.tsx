import { act, renderHook } from '@testing-library/react-hooks';
import React from 'react';
import ReactTestRenderer from 'react-test-renderer';
import ThemingManager, { useTheming } from '..';

describe('components/core/ThemingManager Test Suite', () => {
  it('matches snapshot', () => {
    const jsx = <ThemingManager>ThemingManager Component</ThemingManager>;
    expect(ReactTestRenderer.create(jsx).toJSON()).toMatchSnapshot();
  });

  it('should change theme accordingly', () => {
    const wrapper = ({ children }) => <ThemingManager>{children}</ThemingManager>;
    const { result } = renderHook(() => useTheming(), { wrapper });
    expect(result.current.mode).toBe('light');
    act(() => {
      result.current.toggle();
    });
    expect(result.current.mode).toBe('dark');
  });
});
