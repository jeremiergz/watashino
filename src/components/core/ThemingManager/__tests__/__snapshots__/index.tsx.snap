// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/core/ThemingManager Test Suite matches snapshot 1`] = `"ThemingManager Component"`;
